zip-proportional-output-fee-mechanism-pofm.rst

```diff
@@ -0,0 +1,294 @@
+  transactions, and base_fee = 10000, marginal_fee = 10000 for Transparent transactions 
+  per @nighthawk24's proposal.
+
+
+Security and Privacy considerations
+=============
+
+Non-standard transaction fees may reveal specific users or wallets or wallet versions,
+which would reduce privacy for those specific users and the rest of the network.
+Hence this change should be accepted by a majority of shielded transaction
+software providers before deploying the change.
+
+Long term, the issue of fees needs to be re-visited in separate future
+proposals as the blocks start getting consistently full. New ZIPs with 
+scaling solutions, will need to be evaluated and applied.
+
+
+Denial of Service Vulnerability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A transaction-rate-based denial of service attack occurs when an attacker
+generates enough transactions over a window of time to prevent legitimate
+transactions from being mined, or to hinder syncing blocks for full nodes
+or miners.
+
+There are two primary protections to this kind of attack in Zcash: the
+block size limit, and transaction fees. The block size limit ensures that
+full nodes and miners can keep up with the blockchain even if blocks are
+completely full. However, users sending legitimate transactions may not
+have their transactions confirmed in a timely manner.
+
+Variable fees could mitigate this kind of denial of service: if there are
+more transactions available than can fit into a single block, then a miner
+will typically choose the transactions that pay the highest fees. If
+legitimate wallets were to increase their fees during this condition, the
+attacker would also increase the fees of their transactions. It is
+sometimes argued that this would impose a cost to the attacker that would
+limit the time window for which they can continue the attack. However, there
+is little evidence that the actual costs involved would be a sufficient
+disincentive.
+
+This proposal does not alter how fees are paid from transactions to miners.
+
+Wallet developers and operators should monitor the Zcash network for rapid
+growth in transaction rates.
+
+
+Transaction relaying
+--------------------
+
+zcashd, and potentially other node implementations, implement fee-based
+restrictions on relaying of mempool transactions. Nodes that normally relay
+transactions are expected to do so for transactions that pay at least the
+conventional fee, unless there are other reasons not to do so for robustness
+or denial-of-service mitigation.
+
+
+Mempool size limiting
+---------------------
+
+zcashd and Zebra limit the size of the mempool as described in [#zip-0401]_. This
+specifies a *low\_fee\_penalty* that is added to the "eviction weight" if the
+transaction pays a fee less than the `min_fee` specified by this ZIP.
+
+
+Deployment
+=============
+The height for coordinating deployment is block 1,800,000 for Mainnet, or immediately on 
+implementing this ZIP for Testnet.
```

Comment:
For deployment I recommend this sequence:

- deploy the change in fees as quickly as possible in wallets.
- observe the behavior as wallets deploy the change.
- explicitly define transaction relay rules to exclude/drop transactions below the POFM threshold.
- explicitly define "standard" `getblocktemplate` behavior which excludes txns below the fee.
- release the relay and `getblocktemplate` changes in `zcashd` and `zebrad` whenever ready _as long as_ it won't interfere significantly with the majority of observed transactions when released.

This approach doesn't rely on activation heights and is "as soon as it can ship" for wallets and full nodes. However I am explicitly suggesting some time between wallet deployment and full node deployment to observe the impacts for safety.

I also suggest we exclude both miner txn selection and consensus rules changes from this ZIP. Miner transaction selection is up to miners, but my suggestion does alter `getblocktemplate` defaults. My best intuition is that most miners use that default because until recently blocks have not been full so there's been no benefit to customized txn selection. 

However, I also believe if blocks are consistently full, miners will have competitive pressure to select whichever policy maximizes fee revenue, so we should anticipate that. If there were a sufficient number of both fully-transparent and txns with shielded components to choose from, I regretfully predict they would prefer a fully-transparent subset if all txns use POFM strictly, because more can fit into a block, thus more fees in the block.

I recommend we address that issue _later_ to keep ZIP 317 more tightly focused on solving the current acute pain point.